crates/matrix-sdk/src/sliding_sync/client.rs

```diff
@@ -40,6 +45,97 @@ impl Version {
     }
 }
 
+/// A builder for [`Version`].
+#[derive(Clone, Debug)]
+pub enum VersionBuilder {
+    /// Build a [`Version::None`].
+    None,
+
+    /// Build a [`Version::Proxy`].
+    Proxy {
+        /// Coerced URL to the proxy.
+        url: Url,
+    },
+
+    /// Build a [`Version::Native`].
+    Native,
+
+    /// Build a [`Version::Proxy`] by auto-discovering it.
+    ///
+    /// It is available if the server enables it via `.well-known`.
+    DiscoverProxy,
+
+    /// Build a [`Version::Native`] by auto-discovering it.
+    ///
+    /// It is available if the server enables it via `/versions`.
+    DiscoverNative,
+}
+
+impl VersionBuilder {
+    pub(crate) fn needs_get_supported_versions(&self) -> bool {
+        matches!(self, Self::DiscoverNative)
+    }
+
+    /// Build a [`Version`].
+    ///
+    /// It can fail if auto-discovering fails, e.g. if `.well-known`
+    /// or `/versions` do contain invalid data.
+    pub fn build(
+        self,
+        well_known: Option<&discover_homeserver::Response>,
+        versions: Option<&get_supported_versions::Response>,
+    ) -> Result<Version, Error> {
+        Ok(match self {
+            Self::None => Version::None,
+
+            Self::Proxy { url } => Version::Proxy { url },
+
+            Self::Native => Version::Native,
+
+            Self::DiscoverProxy => {
+                let Some(well_known) = well_known else {
+                    return Err(Error::VersionCannotBeDiscovered(
+                        "`.well-known` is `None`".to_owned(),
+                    ));
+                };
+
+                let Some(info) = &well_known.sliding_sync_proxy else {
+                    return Err(Error::VersionCannotBeDiscovered(
+                        "`.well-known` does not contain a `sliding_sync_proxy` entry".to_owned(),
+                    ));
+                };
+
+                let url = Url::parse(&info.url).map_err(|e| {
+                    Error::VersionCannotBeDiscovered(format!(
+                        "`.well-known` contains an invalid `sliding_sync_proxy` entry ({e})"
+                    ))
+                })?;
+
+                Version::Proxy { url }
+            }
+
+            Self::DiscoverNative => {
+                let Some(versions) = versions else {
+                    return Err(Error::VersionCannotBeDiscovered(
+                        "`/versions` is `None`".to_owned(),
+                    ));
+                };
+
+                match versions.unstable_features.get("org.matrix.simplified_msc3575") {
+                    Some(value) if *value => Version::Native,
+                    _ => {
+                        return Err(Error::VersionCannotBeDiscovered(
+                            "`/versions` does not contain `org.matrix.simplified_msc3575` in its \
+                            `unstable_features`"
+                                .to_owned(),
+                        ))
+                    }
```

Comment:
It's a bit far fetched, but if the unstable feature is set to false, the error message will be incorrect. Probably not a big deal.